flow.py

```diff
@@ -57,17 +66,104 @@ def dest(self):
     def dest(self, dest):
         raise AttributError("Cannot modify a flow's destination node.")
 
-    def make_packet(self, id, src, dest, data):
-        """Method makes the packet and triggers the send_packet method for the
-        host if applicable
+    def check_last(self):
+        """Method that checks if all packets have been acknowledged and
+        updating the NetworkSimulator object accordingly
+        """
+        if len(self.unacknowledged) == 0:
+            self.ns.decrement_active_flows()
+
+    def update_flow(a_packet):
+        """Upon receiving an acknowledgement packet, updating the flow's
+        attributes
+
+        Args:
+            a_packet (AcknowledgementPacket): Packet being sent
+                back from host
+        """
+        print("Flow: updated with acknowledgement packet: ", a_packet.id,
+            "And flow", self.flow_id)
+        self.check_last()
+        self.unacknowledged_packets.remove(a_packet.packet_id)
+
+    def time_out(self, packet_id):
+        """Method where sent packet is added to timed_out_packets array if
+        still unacknowledged after a period of time
 
         Args:
-            id (int): The unique id identifying the packet
+            packet (Packet): Packet being added to time_out dictionary
+        """
+        if packet_id in unacknowledged:
+            timed_out_packets.add(packet_id)
+            unacknowledged_packets.remove(packet_id)
+            print("Packet ", packet_id, " timed out in Flow ", self.flow_id)
```

Comment:
when you use a comma as a separator in print statements, it automatically adds the space so now you just have two spaces and it'll look dumb